zaza/openstack/charm_tests/ceph/dashboard/tests.py

```diff
@@ -309,3 +302,133 @@ def test_saml(self):
             verify=self.local_ca_cert,
             allow_redirects=False)
         self.assertEqual(resp.status_code, requests.codes.ok)
+
+    def is_app_deployed(self, app_name) -> bool:
+        """Check if the provided app is deployed in the zaza model."""
+        try:
+            zaza.model.get_application(app_name)
+            return True
+        except KeyError:
+            return False
+
+    def _get_wait_for_dashboard_assert_state(
+            self, state, message_prefix) -> dict:
+        """Generate a assert state for ceph-dashboard charm blocked state."""
+        assert_state = {
+            'ceph-dashboard': {
+                "workload-status": state,
+                "workload-status-message-prefix": message_prefix
+            }
+        }
+        # Telegraf has a non-standard active state message.
+        if self.is_app_deployed('telegraf'):
+            assert_state['telegraf'] = {
+                "workload-status": "active",
+                "workload-status-message-prefix": "Monitoring ceph"
+            }
+
+        return assert_state
+
+    def verify_ssl_config(self, ca_file):
+        """Check if request validates the configured SSL cert."""
+        rcs = collections.defaultdict(list)
+        units = zaza.model.get_units('ceph-mon')
+        for unit in units:
+            req = self._run_request_get(
+                'https://{}:8443'.format(
+                    zaza.model.get_unit_public_address(unit)),
+                verify=ca_file,
+                allow_redirects=False)
+            rcs[req.status_code].append(
+                zaza.model.get_unit_public_address(unit)
+            )
+        self.assertEqual(len(rcs[requests.codes.ok]), 1)
+        self.assertEqual(len(rcs[requests.codes.see_other]), len(units) - 1)
+
+    def _get_dashboard_hostnames_sans(self):
+        """Get a tuple of hostnames for the units of the application."""
+        ret_val = ['ceph-dashboard']  # Include hostname in san as well.
+        for unit in zaza.model.get_units('ceph-mon'):
+            addr = zaza.model.get_unit_public_address(unit)
+            if addr:
+                ret_val.append(addr)
+        return tuple(ret_val)
+
```

Comment:
Okay; so this function's return value is *very* odd to me.  It appears to be returning a variable-length `tuple()`, and I'm really struggling to work out the justification for this.  AFAICT, this function.

1. Creates a variable length list via `ret_val`
2. Consumes the variable length list int a `tuple` (which essentially takes an iterator), thus returning the tuple and *throwing away* the list structure.
3. Is immediately used by `test_006_charm_config_ssl` which using `*var` which consumes an iterator, which obviously the `tuple` provides, and then throws the tuple away.

So why create the `tuple`?

If you are concerned about immutability, then just make `_get_dashboard_hosnames_sans` a generator (although that can only be used on, or just return the list.

Reply:
Thanks @ajkavanagh, As per the suggestion I have switched to using a generator.